### PR TITLE
Refactored frontend by creating four placeholder extensions

### PR DESCRIPTION
**Pull request by acu192**
The behavior of the old front-end code is retained (i.e. it still queries the localhost GraphQL server for the datasets and shows the JSONified string response in a simple widget (@xmnlab original code)).

This PR takes a stab at organizing the code into what I perceive as the "JupyterLab way". It creates multiple extensions that depend on one another, thus allows for the Phosphor dependency injection system to initialize everything in the proper order, and to pass around interfaces rather than concrete types.

The extensions I created are just placeholders. Their implementations are still `TODO`. But hopefully this gives us a vague outline to go by. @saulshanabrook can judge if this looks okay, compared to how other core extensions are organized.

The four extension are described briefly below (see `frontend/src/index.ts`):

### JupyterLabPlugin&lt;IMetadataApolloGraphQlConnection&gt;

Provides an interface to query the QraphQL server. Requires nothing.

### JupyterLabPlugin&lt;IMetadataCommentsService&gt;

Provides an interface to query/create comments. Requires the `IMetadataApolloGraphQlConnection`. (Hides the GraphQL interface under this higher-level "comment" interface.)

### JupyterLabPlugin&lt;IMetadataDatasetsService&gt;

Provides an interface to query/create dataset metadata. Requires the `IMetadataApolloGraphQlConnection`. 
(Again, hides the GraphQL interface under this higher-level "datasets" interface.)

### JupyterLabPlugin&lt;void&gt;   // UI

Provides an extension to hold the UI of the metadata explorer. Requires `IMetadataCommentsService` and `IMetadataDatasetsService`.

### CODE REVIEW

**Review by xmnlab — APPROVED** [review]
LGTM! thanks @acu192 !
waiting for feeback from @saulshanabrook to merge this PR.

**Comment by acu192** [comment]
One obvious change that's needed is to have the interfaces take callbacks. Then we can hide away the async handling code inside these implementations and not force the UI code to deal with that. But I'd say let's get this merged so that our codebases are aligned again, then we can continue to work toward the correct interface.

File: frontend/src/metadata_concrete/apollo_connection.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  implements IMetadataApolloGraphQlConnection {
+  client: ApolloClient<{}>;
+
+  constructor() {
+    this.client = new ApolloClient({
+      uri: "http://localhost:4000" // TODO: parameterize this
+    });
+  }
+
+  executeQuery(query: object): object {
```
Comment (saulshanabrook) [inline frontend/src/metadata_concrete/apollo_connection.ts:30]: I think this should return a promise.

**Review by saulshanabrook — APPROVED** [review]
Looks good from my end. Let's merge this in and iterate.

**Comment by saulshanabrook** [comment]
> One obvious change that's needed is to have the interfaces take callbacks. Then we can hide away the async handling code inside these implementations and not force the UI code to deal with that. 

I think having the get and update functions all return promises makes sense. We use promises in JupyterLab instead of callbacks for async functions, if I am understanding your suggestion correctly. 

My current thinking is that the datasets in the backend all should have URLs. This is the join key to the frontend dataset registry. i.e. If I choose a CSV file as a dataset, it should have a relative path URL. Then, the graphql backend can store metadata about that dataset, and the URL is the reference to it, so when the frontend sees that URL, it knows "Oh this dataset lives in a file, I can go open that file to access it".

So this looks good from that end! Seems to be compatible with that approach.

**Pull request merged**